misk/src/main/kotlin/misk/web/interceptors/ConcurrencyLimitsInterceptor.kt

```diff
@@ -0,0 +1,79 @@
+package misk.web.interceptors
+
+import com.netflix.concurrency.limits.Limiter
+import com.netflix.concurrency.limits.limiter.SimpleLimiter
+import misk.Action
+import misk.exceptions.StatusCode
+import misk.logging.getLogger
+import misk.web.AvailableWhenDegraded
+import misk.web.ConcurrencyLimitsOptIn
+import misk.web.NetworkChain
+import misk.web.NetworkInterceptor
+import java.time.Clock
+import javax.inject.Inject
+import javax.inject.Singleton
+import kotlin.reflect.full.findAnnotation
+
+/**
+ * Detects degraded behavior and sheds requests accordingly. Internally this uses adaptive limiting
+ * as implemented by Netflix's [concurrency-limits][concurrency_limits] library. See
+ * [AvailableWhenDegraded] for further documentation.
+ *
+ * [concurrency_limits]: https://github.com/Netflix/concurrency-limits/
+ */
+internal class ConcurrencyLimitsInterceptor internal constructor(
+  private val actionName: String,
+  private val limiter: Limiter<String>
+) : NetworkInterceptor {
+
+  override fun intercept(chain: NetworkChain) {
+    val listener: Limiter.Listener? = limiter.acquire(actionName).orElse(null)
+
+    if (listener == null) {
+      logger.error { "concurrency limits interceptor shedding $actionName" }
+      chain.httpCall.statusCode = StatusCode.SERVICE_UNAVAILABLE.code
```

Comment:
We want to use 429
https://httpstatuses.com/429

Reply:
Really hard to say if it’s client’s fault or server’s fault that the server is overloaded. Franklin does 503 here. Feel strongly on this?

Reply:
(I’m going to merge with 503 to get things unstuck, and I’m totally open to changing it to 429.)

Reply:
We went with 503 because we were nervous about leaking load problems to the outside world. I think ideally we use 429 and Envoy scrubs on the way back out to the internet

Reply:
lol "leaking load problems". They'll never know!

Reply:
I’m going to leave this as-is for now.

Reply:
The nice thing with using 429 is that it signals to the clients that they should back off. For example there is a method to call on Limiter.Listener when the client sends 429 which makes the limiter decrease rate significantly:
https://github.com/Netflix/concurrency-limits/blob/master/concurrency-limits-core/src/main/java/com/netflix/concurrency/limits/Limiter.java#L45

I think unfortunately that means we do have to announce this to the world because ideally we would also want our mobile app clients to implement back off as well in these situations. If Envoy scrubs it that's no good.

Reply:
But I'm fine with merging this as is if we can keep the discussion going. What's so bad about announcing over capacity to the world?

Reply:
In theory I agree with @tirsen, 429 is more appropriate. In reality, probably doesn't really matter. Clients should really back off on all errors, including 5xxs.